src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD112AvoidReturningNullTaskAnalyzer.cs

```diff
@@ -0,0 +1,65 @@
+﻿namespace Microsoft.VisualStudio.Threading.Analyzers
+{
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using Microsoft.CodeAnalysis.Operations;
+    using Microsoft.VisualStudio.Threading.Analyzers.Lightup;
+
+    /// <summary>
+    /// Finds await expressions on <see cref="Task"/> that do not use <see cref="Task.ConfigureAwait(bool)"/>.
+    /// Also works on <see cref="ValueTask"/>.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class VSTHRD112AvoidReturningNullTaskAnalyzer : DiagnosticAnalyzer
+    {
+        public const string Id = "VSTHRD112";
+
+        internal static readonly DiagnosticDescriptor Descriptor = new DiagnosticDescriptor(
+            id: Id,
+            title: new LocalizableResourceString(nameof(Strings.VSTHRD112_Title), Strings.ResourceManager, typeof(Strings)),
+            messageFormat: new LocalizableResourceString(nameof(Strings.VSTHRD112_MessageFormat), Strings.ResourceManager, typeof(Strings)),
+            helpLinkUri: Utils.GetHelpLink(Id),
+            category: "Usage",
+            defaultSeverity: DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+
+        /// <inheritdoc />
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterOperationBlockStartAction(Utils.DebuggableWrapper(context => AnalyzeOperationBlockStart(context)));
+        }
+
+        private static void AnalyzeOperationBlockStart(OperationBlockStartAnalysisContext context)
+        {
+            if (context.OwningSymbol is IMethodSymbol method &&
+                !method.IsAsync &&
+                Utils.IsTask(method.ReturnType))
+            {
+                context.RegisterOperationAction(Utils.DebuggableWrapper(context => AnalyzerReturnOperation(context)), OperationKind.Return);
+            }
+        }
+
+        private static void AnalyzerReturnOperation(OperationAnalysisContext context)
+        {
+            var returnOperation = (IReturnOperation)context.Operation;
+
+            if (returnOperation.ReturnedValue != null && // could be null for implicit returns
+                returnOperation.ReturnedValue.ConstantValue.HasValue &&
+                returnOperation.ReturnedValue.ConstantValue.Value == null &&
+                returnOperation.ReturnedValue.Syntax is { } returnedValueSyntax &&
+                !context.Compilation.GetSemanticModel(returnedValueSyntax.SyntaxTree).GetNullableContext(returnedValueSyntax.SpanStart).AnnotationsEnabled())
```

Comment:
I did something similar on roslyn-analyzers and I end-up with a `RS1030` so I think we shouldn't be moving forward like this. @sharwell any idea of how I could rework this? On https://github.com/dotnet/roslyn-analyzers/pull/3352 we decided to rewrite the analyzer as syntax analyzer but that seems more complex in here.

Reply:
@sharwell @AArnott I have tried a couple of things and I cannot find any good way (except for a full node analyzer) to get the semantic model using the operation analyzer. I am not sure how to move forward:

1/ Revert the part trying to handle the `#nullable enable` part.

2/ Keep current implementation. That's probably a bad idea as an analyzer was written to prevent this case.

3/ Move to a syntax analyzer.

4/ Any other way?

Reply:
Personally I think we can drop no. 2.

Reply:
It's ok when bumping the version of roslyn we are using.